`WCProductReviewModel` to `Room`

wzieba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: AINFRA-129
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR migrates `WCProductReviewModel` from WellSql to Room.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

Add reviews to a product on web and verify the review is available in the app as well in 2 flows: 
- Product Details -> Reviews
- Menu -> Reviews

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
<img width="1600" alt="Screenshot 2025-05-20 at 10 07 58" src="https://github.com/user-attachments/assets/9be5ba22-5117-4d5b-85ab-846fdf8d2efb" />


- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Refactor WCProductReviewModel to a data class

wzieba committed:
Introduce `ProductReviewsDao`

`:libs:fluxc-plugin:assembleDebug` passes

wzieba committed:
Add database migration. Move all product reviews database tests from SqlUtilsTest. Adjust unit tests to execute

`:libs:fluxc-plugin:testDebugUnitTest` passes

wzieba committed:
Adjust `WooCommerce` module codebase and tests to migrated WCProductReviewModel

wzieba committed:
Improve `ProductReviewsDaoTest`

Remove repetitions, simplify tests, split to smaller and focused ones

wzieba committed:
Fix styling

wzieba committed:
Drop `WCProductReviewModel` tabel from WellSql

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Fix `:libs:fluxc-tests` tests

wzieba committed:
Inject `ProductReviewsDao` in DI graph

ParaskP7 commented:
FYI: I am currently testing how reviews are working on `trunk` and prior to this migration. It is very interesting to me that reviews are being constantly recreated, for example, I have 5 reviews in total and: Cc @wzieba
1. When I go the `Menu` -> `Reviews`, those 5 reviews are added to `wp-fluxc.WCProductReviewModel`. ✅
2. Now, when I go back to `Menu`, and then enter `Review` again, the previous 5 reviews are deleted and new 5 reviews are added (I am seeing the `_id` changing). 🤔
3. Also, when I go back to `Products` -> `Product` with 1 review -> `Reviews`, the previous 5 reviews are deleted and one new review is added. 🤔

ParaskP7 commented:
> FYI: I am currently test how reviews are working on `trunk` and prior to this migration. It is very interesting to me that reviews are being constantly recreated, for example, I have 5 reviews in total and: Cc @wzieba

Same thing happens now that I am testing the migration, although the "good" thing about `wc-android-database.ProductReviewEntity` is the fact that there is no `_id` anymore, so this "problem" is not that visible... 🤷

PS: Another, I guess unrelated, thing I noticed is that whenever I look at a product, even if that product have "approved" reviews, it is showing `no approved reviews`. Maybe something for the product team to check-on. 🤔

ParaskP7 commented:
> FYI: I am currently test how reviews are working on trunk and prior to this migration. It is very interesting to me that reviews are being constantly recreated, for example, I have 5 reviews in total and: Cc @wzieba

Ok, [this](https://github.com/woocommerce/woocommerce-android/blob/trunk/libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCProductStore.kt#L1354-L1356) here answers my own question... ✅ 

> PS: Another, I guess unrelated, thing I noticed is that whenever I look at a product, even if that product have "approved" reviews, it is showing no approved reviews. Maybe something for the product team to check-on. 🤔

Still interested in that above though... 💭

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/WCProductReviewModel.kt:
```diff
@@ -1,75 +1,71 @@
+// todo: as soon as SiteModel is migrated to Room, add foreign key constraint
+@Entity(
+    tableName = "ProductReviewEntity",
+    primaryKeys = ["localSiteId", "remoteProductReviewId"],
```
Question (❓): Shouldn't `remoteProductId` be included in that primary key combination, the `UNIQUE` constraint from before the migration was indeed using it? 🤔

wzieba replied:
Sure 👍 I don't think it's necessary (I think remote id of a review is unique in context of a single page) but nevertheless, this PR shouldn't change behaviour. Thanks for noticing, addressed in: be1dad4390369f33053bb456f130bdd376725161

ParaskP7 replied:
Awesome @wzieba , thanks for addressing this and updating the schema, LGTM! 🙇 ❤️ 🚀 

> ...this PR shouldn't change behaviour.

💯

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/WCProductReviewModel.kt:
```diff
@@ -1,75 +1,71 @@
-@RawConstraints(
-        "FOREIGN KEY(LOCAL_SITE_ID) REFERENCES SiteModel(_id) ON DELETE CASCADE",
-        "UNIQUE (REMOTE_PRODUCT_REVIEW_ID, REMOTE_PRODUCT_ID, LOCAL_SITE_ID) ON CONFLICT REPLACE"
+// todo: as soon as SiteModel is migrated to Room, add foreign key constraint
```
👍

ParaskP7 commented on libs/fluxc-plugin/src/test/java/org/wordpress/android/fluxc/persistence/dao/ProductReviewsDaoTest.kt:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    @Test
+    @Ignore("This test is ignored until SiteModel is moved to Room and foreign key constraints are added")
```
👍

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/product/ProductRestClient.kt:
```diff
@@ -2128,18 +2124,18 @@ class ProductRestClient @Inject constructor(
     }
 
     private fun productReviewResponseToProductReviewModel(response: ProductReviewApiResponse): WCProductReviewModel {
```
Minor (🔍): Consider adding the `localId: LocalId` parameter to this private function and using `site.localId()` on every such 3 x usage, that is, instead of the copy function.

wzieba replied:
Good idea 👍 3f14a34ce0c5890f3f61e23ee1e374e654e4fff4

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/ProductReviewsDao.kt:
```diff
@@ -0,0 +1,67 @@
+    abstract suspend fun upsertProductReviews(productReviews: List<WCProductReviewModel>)
+
+    @Delete
+    abstract suspend fun deleteProductReview(productReview: WCProductReviewModel)
```
Very Minor (🔍): Consider moving these 3 x upsert/delete db function at the bottom, just for us to be consistent with the rest of the daos.

ParaskP7 commented on libs/fluxc-plugin/src/test/java/org/wordpress/android/fluxc/persistence/dao/ProductReviewsDaoTest.kt:
```diff
@@ -0,0 +1,158 @@
+
+    @Test
+    fun testUpdateProductReview() = runTest {
+        val review = getProductReviews(site.id)[0]
```
Question (❓): For this test to differ from the above, shouldn't we first insert this `review` to db, before calling `upsertProductReview(...)` on the `updatedReview`? 🤔

```kotlin
sut.upsertProductReview(review)
```

wzieba replied:
Good point, I forgot about it. Addressed in 5ff11bb1f9bec510197608f91e8e200a3168d18e

ParaskP7 commented on libs/fluxc-plugin/src/test/java/org/wordpress/android/fluxc/persistence/dao/ProductReviewsDaoTest.kt:
```diff
@@ -0,0 +1,158 @@
+        sut.upsertProductReviews(reviews)
+
+        // Verify products inserted
+
```
Super Minor (🔍): Consider removing the extra empty line.

ParaskP7 reviewed (commented):
👋 @wzieba !

I have reviewed and tested this PR as per the instructions, everything looks to be working as expected, good job! 🌟 

-----

I have left a couple of questions (❓) and some minor (🔍) comments for you to consider. I am **NOT** going to approve this PR for now, but also, not going to block it either, I just want to discuss this primary key question (❓) first.

wzieba committed:
Fix update product review unit test

https://github.com/woocommerce/woocommerce-android/pull/14091/files/3e39d81a761e9f04e51fba28d04cd202558477bf#r2097749932

wzieba committed:
Simplify `productReviewResponseToProductReviewModel`

https://github.com/woocommerce/woocommerce-android/pull/14091/files/3e39d81a761e9f04e51fba28d04cd202558477bf#r2097744354

wzieba committed:
Include `remoteProductId` to the primary key of `ProductReviewEntity`

It shouldn't be needed (I think remote ids of reviews are unique in context of a site) but nevertheless, it's good to keep the previous behaviour untouched

ParaskP7 commented:
Everything LGTM now, ✅ this to unblock, thanks for all the extra changes @wzieba ! 🙇 ❤️ 🚀

wzieba committed:
Merge branch 'trunk' into migrate_product_review_model
